Merge in latest changes from dev/gfdl

uwagura opened the pull request:
As titled.

Hallberg-NOAA committed:
*Update defaults for EQN_OF_STATE and 14 parameters

  Updated the default values of 15 runtime parameters, as agreed upon in a MOM6
consortium wide conversations on July 29, 2024 and December 16, 2024.  The most
prominent of these is that the default equation of state is now EQN_OF_STATE =
"WRIGHT_FULL", in place of the buggy previous default of "WRIGHT".

  The 8 answer date parameters REGRIDDING_ANSWER_DATE, TIDES_ANSWER_DATE,
WAVE_INTERFACE_ANSWER_DATE, MEKE_GM_SRC_ANSWER_DATE, NDIFF_ANSWER_DATE,
KPP%ANSWER_DATE, HOR_DIFF_ANSWER_DATE and LOTW_BBL_ANSWER_DATE all now take
their default values from DEFAULT_ANSWER_DATE.

  The bug-retention parameters MEKE_GM_SRC_ALT_SLOPE_BUG, HOR_DIFF_LIMIT_BUG,
BACKSCATTER_UNDERBOUND, DETERMINE_TEMP_CONVERGENCE_BUG, LA_MISALIGNMENT_BUG and
IDL_HURR_SCM_EDGE_TAPER_BUG are now false by default.

  The MOM_input files for the test cases in the `.testing/tc[01234]` directories
were updated to explicitly set all of these parameters to their previous default
values if they are used in the relevant cases, as well as parameters that will
be change following discussions from June 2, 2205.

  Bitwise identical answers are recovered if all 15 of these parameters are set
explicitly, but answers can change if any of these parameters take default
values.  All of these default changes are the consensus decision of
consortium-wise MOM6 dev calls.

awallcraft committed:
Add ALE_COORDINATE_CONFIG HYBRID_3D and HYBRID_MAP

Add two options supporting 3D HYCOM1 TARGET_DENSITIES and ALE_RESOLUTION.

The fully general interface is ALE_COORDINATE_CONFIG=HYBRID_3D which
is like HYBRID except that the netCDF variables (e.g. sigma2 and dz)
are 3D fields (x-y-z).  Like HYBRID, the 2nd variable (e.g. dz) can be
replaced by a FNC1 string that sets a dz profile that is used everywhere.

The typical use of 3D HYCOM1 fields is to have different TARGET_DENSITIES
in semi-enclosed seas for layers that are always below its sill depth.
For example, in the OM4 75-layer setup layers 71 amd 72 are only active
in the Arctic and layers 74 and 75 are only active in the Mediterranean.
So the number of layers can be reduced to 71 by specifying different
deep targets in these semi-emclosed seas.  In this case ALE_RESOLUTION
would be the same everywhere, but the resulting 71 layer configuration
can be further improved by using shallow ALE_RESOLUTION in the northern
hemisphere and the standard, deep, dz's in the southern hemisphere.
This is safe, because deep layers are never in fixed coordinates near
the equator.

To simplify this typical use case, ALE_COORDINATE_CONFIG=HYBRID_MAP
requires three netCDF variables (e.g. map,sigma2,dz) with the last
two being 2-D z-index fields containing a small number of profiles
and the first a 2-D x-y field of index values indicating which profile
to use at each location.  Like HYBRID and HYBRID_3D, the 3rd variable
(e.g. dz) can be replaced by a FNC1 string that sets a dz profile
that is used everywhere.

Answers are not changes unless ALE_COORDINATE_CONFIG is set to HYBRID_3D
or HYBRID_MAP.

awallcraft committed:
Set the halos of new 3D arrays

awallcraft committed:
Add SHALLOW_ALE_RESOLUTION

SHALLOW_ALE_RESOLUTION implements a HYBGEN-style Z-sigma-Z near surface
fixed coordinate for HYCOM1.  For example the US Navy's GOFS 3.1 HYCOM
setup has 41 layers, with the top 14 layers in a Z-sigma-Z configuration.
For MOM6 HYCOM1 this is: SHALLOW_ALE_RESOLUTION = 14*1.0,27*0.0 for 14
1m "shallow" layers.

Let N_SIGMA be the number of consecutive non-zero entries, typically < NK.
When rest depth is shallower than SUM(SHALLOW_ALE_RESOLUTION(1:N_SIGMA))
use SHALLOW_ALE_RESOLUTION.  When rest depth is deeper than
SUM(SHALLOW_ALE_RESOLUTION(1:N_SIGMA)) use ALE_RESOLUTION.  Otherwise
use a linear sum of the two weighted by rest depth.

The default of all zeros turns this option off, and when off answers are
unchanged.  The new parmeter SHALLOW_ALE_RESOLUTION is only present when
using HYCOM1.

Hallberg-NOAA committed:
*Update defaults for 6 parameters (VISC_REM_BUG, )

  Updated the default values of 6 runtime parameters, as agreed upon in a MOM6
consortium wide conversations on June 2, 2025. VISC_REM_BUG and FRICTWORK_BUG
are now false by default. MASS_WEIGHT_IN_PRESSURE_GRADIENT_TOP now takes its
default value from the setting for MASS_WEIGHT_IN_PRESSURE_GRADIENT.  Similarly,
the default for DRAG_DIFFUSIVITY_ANSWER_DATE is now set to follow
SET_DIFF_ANSWER_DATE. KELVIN_WAVE_VEL_NUDGING_TIMESCALE is now set to a negative
value by default, forcing the user to provide a valid value at runtime.
SHELFWAVE_CORRECT_AMPLITUDE now defaults to True.

Hallberg-NOAA committed:
+Obsoleted 7 parameters, including BETTER_BOUND_KH

  Obsoleted 7 runtime parameters that always take the same value, as agreed upon
in a MOM6 consortium wide conversations on June 2, 2025, and added tests to
catch these parameters in MOM_obsolete_params.  The parameters that were
obsoleted include BETTER_BOUND_KH, BETTER_BOUND_AH, USE_DIABATIC_TIME_BUG,
FIX_UNSPLIT_DT_VISC_BUG, CFL_BASED_TRUNCATIONS and KD_BACKGROUND_VIA_KDML_BUG.
The runtime parameter MAXVEL still exists, but it is logged in a different order
than before, changing the contents of the MOM_parameter_doc files. All answers
are bitwise identical in cases that run, but cases that use these parameters
will experience a fatal error.

Hallberg-NOAA committed:
+Add OBC_REVERSE_SEGMENT_ORDER

  Added the new debugging runtime parameter OBC_REVERSE_SEGMENT_ORDER that can
be set to true to internally store and apply the OBC segments in the reverse of
the order with which they are specified by the used.  When the OBC segments are
order independent, changing this value should give bitwise identical results.
There is a new integer argument indicating the segment to use for setting
parameters in setup_u_point_obc and setup_v_point_obc, but these are internal
routines that are not publicly visible.  This commit also changes some of the
segment index variables from l (which can be hard to distinguish from 1) to n,
and it standardizes the spacing in some loops over segments or in variable
assignments to better align with the MOM6 style guide.  By default the new
runtime parameter is false and all answers are bitwise identical.

Hallberg-NOAA committed:
+Use OBCs to set masks

  Added the ability to take open boundary conditions into account when setting
up the masks, by providing the new optional arguments OBC_dir_u and OBC_dir_v in
calls to initialize_masks().  The new optional argument open_corner_OBCs to
initialize_masks() can be used to specify that convex corners between OBC
segments should be treated as open.  These had to be handled via optional
arguments because SIS2 also calls initialize_masks and it is under separate
version control.  By default all answers are bitwise identical, but there are
new optional arguments to a publicly visible routine.

Hallberg-NOAA committed:
+Optionally avoid projecting data for OBCs

  Adds the option to avoid using exterior data at open boundary condition
points, and to avoid projecting information outside across open boundary
condition faces.  Such projections are problematic because there can be pairs
of OBCs that would project inconsistent values into the same point.  This
option is enabled by setting the new runtime parameter OBC_PROJECTION_BUG =
False. To implement these changes, the weights for the barotropic solver to use
when interpolating total thickness onto vorticity points when calculating the
barotropic potential vorticity for use in the Coriolis scheme are now being
stored in the new q_wt element in the barotropic control structure.  These
weights are set up during initialization with information about open boundary
conditions optionally taken into account.  In addition the total thicknesses at
open boundary velocity points that are combined with the potential vorticities
to give the barotropic Coriolis terms are always taken from the interior of an
OBC face.  Some inconsistencies in the units of a negligible but positive cell
volume in the denominator of the barotropic PVs were also corrected in a way
that will not change answers when dimensional rescaling is not being applied.

  This commit also includes the run-time configurable option to set the corner
point land-mask at the bay-like intersection of open boundary conditions to be
unmasked so that open boundary conditions can be applied there.

  By default all answers are bitwise identical, but answers in cases with open
boundary conditions do change when these new capabilities are enabled.

marshallward committed:
makedep: Update interpreter directive to python3

The interpreter directive ("shebang") of makedep is updated to
`python3`, rather than the version-agnostic `python`.

Although we never invoke the shebang of the script, there are OS
environments out there which will object to any presence of a
versionless python.  PEP 394 also strongly recommends the adoption of
python3 as the executable name, regardless of Py2 support.

Hallberg-NOAA committed:
Set intent(out) stub MOM_generic_tracer arguments

  Added code to set all intent(out) arguments in the stub version of the
MOM_generic_tracer routines MOM_generic_tracer_stock(),
MOM_generic_tracer_min_max() and MOM_generic_tracer_get() to avoid compile-time
warnings.  These stub routines are never supposed to be used, and there would be
a fatal error during the registration phase if they are used.  The values that
are set in these stub routines are extremely large or otherwise nonsensical so
that it would make it easier to debug them if they were used accidentally.  All
answers are bitwise identical in any cases that run successfully.

marshallward committed:
Merge pull request #1670 from Hallberg-NOAA/update_param_defaults_2025

+*Update parameter defaults for 2024 and 2025, including EQN_OF_STATE

marshallward committed:
Merge branch 'main' into dev/gfdl

Hallberg-NOAA committed:
+Separate OBC_for_remap and OBC_for_bug arguments

  Add separate OBC_for_remap and OBC_for_bug arguments to MOM_initialize_state()
to clarify the distinct roles of the two arguments, to make it clear that one
can not be used without rotation, and to prepare for the eventual obsoleting of
OBC_RESERVOIR_INIT_BUG.  This commit also simplifies the logic setting ntstep
when do_thermo is false. All answers are bitwise identical, but there is a new
optional argument to MOM_initialize_state and another argument has been renamed.

Hallberg-NOAA committed:
+Add chksum_OBC_segment_data

  Added the new subroutine chksum_OBC_segment_data() from the contents of the
loop over the segments in chksum_OBC_segments, as this can be useful when called
separately for debugging some of the constructs that will be coming in
subsequent commits.  Also added rotate_OBC_segment_values_needed() to copy over
the values_needed and _index variables between segments, and to avoid the need
for duplicate code blocks in rotate_OBC_segment_data() and rotate_OBC_config().
There is also some minor revisions to rename segnam to segname for greater
clarity.  This commit also makes greater use of the I0 format that was
introduced with Fortran 95 to simplify or shorten some error messages.  All
answers are bitwise identical but there are two new internal subroutines in the
MOM_open_boundary module.

Hallberg-NOAA committed:
+Call initialize_segment_data with rotated OBCs

  Call initialize_segment_data() with the rotated OBC type.  This includes the
addition of a new turns arguments to initialize_segment_data() and the
elimination of the ocean_grid_type argument to this routine.  This includes the
addition of the new function rotated_field_name() to swap the names of u- and v-
velocity fields when the grid is rotated, avoiding the duplication of code.
Internally in initialize_segment_data(), num_fields was renamed to reflect the
fact that it is actually the number of tracer or other fields that are
initialized via the OBC manifest string (which does not include generic OBGC
tracers), and not the total number of fields on a segment.  Code was also added
to allow for rotate_OBC_config() to be called either before or after
initialize_segment_data().   With these changes, the call to
initialize_segment_data() can be moved much later in the order of the
initialization calls to address issues with the initialization of OBGC tracers
with open boundary conditions.  All answers are bitwise identical, but there are
changes to the arguments of the publicly visible routine
initialize_segment_data().

Hallberg-NOAA committed:
+Add RESCALE_STRONG_DRAG

  Added the new runtime option RESCALE_STRONG_DRAG, that can be set to true to
reduce the barotropic contribution to the layer accelerations to account for the
difference between the forces that can be counteracted by the stronger drag with
BT_STRONG_DRAG and the average of the layer viscous remnants after a baroclinic
timestep.  In testing, this new capability eliminates some of the growing
instabilities that can occur with an ice shelf and BT_STRONG_DRAG set to true.
This commit also adds new diagnostics of the barotropic step viscous
remnants and the eta anomalies contributing to barotropic pressure forces,
either averaged over the barotropic step or at each barotropic step.  By
default all answers are bitwise identical, but there is a new runtime parameter
and 4 new diagnostics.

awallcraft committed:
Non-integer HYBRID_MAP values

The 2-d REAL map array in HYBRID_MAP usually contains integer values
each referencing one profile.  It can instead contain non-integer
values of the form I+frac, which indicate a weighted sum of profiles:
(1-frac) p(I) + (frac) p(I+1).  The same profile can be used multiple
times, e.g. if 1st profile is also 4th can get profiles between 1 and 2
and between 1 and 3.

HYBRID_3D is more general, but HYBRID_MAP covers most practical uses.

Hallberg-NOAA committed:
+Restore open_boundary_setup_vert

  Returned the functionality of open_boundary_setup_vert() to a separate
routine, and no longer have initialize_segment_data() carry out these
initialization steps.  These routines had previously been combined in the
expectation that they would always occur at about the same point in the code
but this turns out not to be the case for generic tracers with open boundary
conditions.  This renewed routine is called directly form initialize_MOM.  All
answers are bitwise identical but there is one added public OBC interface.

Hallberg-NOAA committed:
*Call initialize_segment_data after call_tracer_register_obc_segments

  Move call to initialize_segment_data() after the call to
call_tracer_register_obc_segments() to correct an answer changing bug in models
with OBGC tracers and open boundaries that was causing the dev/CEFI branch
answers not to reproduce with dev/gfdl.  This commit will change answers (and
restore previous answers) in cases that use open boundary conditions with
biogeochemical tracers, such as those in COBALT.

Hallberg-NOAA committed:
Eliminate rotate_OBC_segment_data

  The temporary blocks of code allowing for initialize_segment_data either
before or after rotate_OBC_config() were eliminated, as was the now unused
subroutine rotate_OBC_segment_data().  All of the functionality of
rotate_OBC_segment_data had already been merged into initialize_segment_data.
All answers are bitwise identical, but there is one less internal subroutine,
and the order of calls to initialize the open boundary conditions is now less
flexible.

Wendazhang33 committed:
Added the WENO schemes to reconstruct PV in MOM_CoriolisAdv (#946)

* Added a 7th-, 5th- and 3rd-order WENO schemes to reconstruct the PV and added the Koren scheme to reconstruct the kinetic energy in MOM_CoriolisAdv:

- Set CORIOLIS_SCHEME = "WENOVI7TH_PV_ENSTRO", "WENOVI5TH_PV_ENSTRO", :WENOVI3RD_PV_ENSTRO" to use a 7th-, 5th-, and 3rd-order WENO scheme, respectively, to reconstruct PV.
- Set KE_SCHEME = "KE_UP3" to use a third-order upwind scheme with the Koren flux limiter to reconstruct the kinetic energy. Set KE_USE_LIMITER = False (default is True) to not use the flux limiter. The KE_UP3 scheme is recommended to be used along with the WENO Coriolis schemes.
- Set WENO_VELOCITY_SMOOTH = True to use the velocity field to estimate the smoothness indicator for WENO. This scheme tends to yield higher-order of accuracy and thus more energetic eddy fields.
- The WENO schemes conduct a thickness-weighted reconstruction for PV to avoid PV singularities.
- The order of WENO is reduced to 5th, 3rd, and 1st subsequently when it approaches lateral boundaries
- The halo size is 4 and 5 for the 7th- and 5th-order WENO schemes. When they are used, parameters USE_WENO and WENO_stencil are passed to MOM_dynamics_split_RK2 to increase the halo update size. Only MOM_dynamics_split_RK2 is changed now. The other RK2 modules will need to be updated later.

* Replaced the WENO_stencil by the CorAdv_stencil function that is passed to MOM_dynamics_split_RK2

* Debug

* Replaced a few ratios to variables. Corrected the unit of some variables

* Changed adding area_tiny and h_tiny to taking the max

* Removed the dimensional constant in WENO weighting

* Added stencil in $OMP(shared). Added additional comments

* Removed trailing space

* Changed the zero check if statement for fac_fn. Changed Jsq, Jeq, Isq, Ieq to private variables

* Addressed couple of rotational symmetry issues

* Removed commented lines in RK2 and added WENO stencil information to RK2B

* Added Is_q, Ie_q, Js_q, Je_q to set bounds for Coriolis schemes

* Enlarged halo update size for WENO schemes in dynamic_unsplit modules

* Removed trailing space

* Moved the update of h to a group pass in RK2 module

* Removed unnecessary lines

---------

Co-authored-by: Alistair Adcroft <Alistair.Adcroft@noaa.gov> and Robert Hallberg <Robert.Hallberg@noaa.gov>

awallcraft committed:
indent continuations, source code <= 100 chars

Hallberg-NOAA committed:
Fix calculation of CAv_Stokes diagnostic

  Corrected a horizontal indexing bug in the calculation of the CAv_Stokes
diagnostic, making it rotationally consistent and consistent with the
calculation of CAu_Stokes.  This bug has been there since the CAv_Stokes
diagnostic was originally added.  The loop range over which qS is calculated was
also reduced to the range over which it is used.  All solutions are bitwise
identical, but this commit does change the values of a (perhaps infrequently
used) diagnostic.

breichl committed:
Add option to horizontally homogenize the Stokes drift when used via … (#967)

* Add option to horizontally homogenize the Stokes drift when used via the dataoverride surfbands procedure.

* Add variable description in new method for horizontally averaging Stokes drift.

---------

Co-authored-by: brandon.reichl <brandon.reichl@noaa.gov>

uwagura commented:
The diagnostic regression tests are failing due to a change in masking  that we are just encountering now- see https://github.com/NOAA-GFDL/MOM6/pull/950#issuecomment-3211498183

Since answers answers are not changing, this PR is good to go

theresa-cordero commented:
These commits do not change answers for the NEP10 and NWA12 regression tests on Gaea. 

In order to maintain the answers for the NEP10, the following parameters need to be specified in the `MOM_override` file: 
```
TIDES_ANSWER_DATE = 20230630
MASS_WEIGHT_IN_PRESSURE_GRADIENT_TOP = False
WAVE_INTERFACE_ANSWER_DATE = 20221231 
```

Hallberg-NOAA committed:
+Correct halo update sizes and reduce halo updates (#969)

Added the new argument dyn_h_stencil to initialize_dyn_split_RK2 and the
other 3 dynamic core initialization routines to return the size of the stencil
for thicknesses as used by the dynamic core, depending on the options that are
being used for the Coriolis and continuity schemes, and then used this in a set
of halo updates in step_MOM_dynamics.  With this change some additional halo
updates that have recently been added inside of step_MOM_dyn_split_RK2 and the
other 3 dynamic core time stepping routines could be eliminated.  All answers
are bitwise identical, but there is a new argument to 4 public interfaces.

Co-authored-by: Marshall Ward <marshall.ward@gmail.com>

Hallberg-NOAA committed:
Correct multi-PE velocity truncation counts (#955)

Modified the conditions that determine when to increment the count of velocity
truncations to use the thickness as interpolated to velocity points to determine
whether layers are thick enough to be counted, rather than the arithmetic mean
thickness, and only count truncations that occur in the non-symmetric
computational domain to avoid double counting.  The filtering thicknesses
should be very similar in the ocean interior, but they will differ at open
boundary condition points.  The corrected counting was verified by running the
sloshing/layer test case with a maximum CFL set to 0.01 to create lots of
truncations, and then verifying that the truncation count is now the same on 1
and 10 PEs, whereas before it was not.  All solutions are bitwise identical, but
the reported truncation counts in the ocean.stats files can change in
multiple-PE cases with velocity truncations.

Pperezhogin committed:
ANN parameterization of horizontal momentum eddy fluxes (#944)

* Add MOM_ANN module

* Mesoscale momentum parameterization with ANN

- Computes subgrid stress using ANN in MOM_Zanna_Bolton
- Uses MOM_ANN module for ANN inference

Equivalent MOM_override for defaults
```
USE_ZB2020 = True
ZB2020_USE_ANN = True
USE_CIRCULATION_IN_HORVISC = True
ZB2020_ANN_FILE_TALL = /path/to/ocean3d/subfilter/FGR3/EXP1/model/Tall.nc
```

* Mesoscale momentum parameterization with ANN (#2)

Blank commit after squash/rebase was handled on command line

* Moved MOM_ANN.F90 to src/framework/

* Minor refactor of MOM_ANN

- Removed unused modules
- Removed unused MOM_memory.h
- Added input and output means which default to 0 and
  do not need to be present in the weights file
- Gave defaults to means, norms, tests so that they do
  no need to be present in file
- Added missing array notation "(:)"
- Minor formatting

* Adds unit tests and timing test to MOM_ANN

- Added ANN_allocate, set_layer, set_input_normalization, and
  set_output_normalization methods to allow reconfiguration during
  unit tests
- Added ANN_unit_tests with some simple constructed-by-code
  networks with known solutions
- Added config_src/drivers/unit_tests/test_MOM_ANN.F90 to drive
  unit tests
- Added  config_src/drivers/timing_tests/time_MOM_ANN.F90 as
  rudimentary for timing inference

* Adding multiple forms of inference

- Adds inference operating on array (instead of single vector of
  features)
- Implements several different versions of inference with various
  loop orders
  - Involves storing the transpose of A in the type
  - Tested by checking inference on same inputs is identical between
    variants
    - Added randomizers to assist in unit testing
- Adds timing of variants to config_src/drivers/timing/time_MOM_ANN.F90
- Adds an interface (MOM_apply) to select preferred version of
  inference subroutine
- Added command line args to time_MOM_ANN.F90 to allow more rapid
  evaluation of performance

Variants explored, timed with gfortran (13.2) -O3 on Xeon:
- vector_v1:
  - original inference from Pavel
- vector_v2:
  - allocate work arrays just once, using widest layer
  - loop over layers in 2's to avoid pointer calculations and copies
  - speed up, x0.8 relative to v1
- vector_v3:
  - transpose loops
  - slow down, x1.54 relative to v1
- vector_v4:
  - transpose weights with same loop order as v1
  - slow down, x1.03 relative to v1
- array_v1:
  - same structure as v2, working on x(space,feature) input/outputs
  - speed up, x0.41 relative to v1
- array_v2:
  - as for array_v1 but with transposed loop order
  - apply activation function on vector of first index while in cache
  - speed up, x0.35 relative to v1
- array_v3:
  - same structure as v2, working on x(feature,space) input/outputs
  - speed up, x0.58 relative to v1

* Renamed ANN variants and added some module documentation

- Added module dox
- Renamed _v1, _v2 etc to labels
- Added ANN_apply_array_sio to ANN_apply interface
- Replaced "flops" with "MBps" in timing output

* Removed alternative variants of ANN in favor of optimized

- Deleted variants of ANN that did not perform as well as the two
  versions that remain.

* Apply array_sio function in ANN inference for momentum fluxes (#5)

* Apply array_sio ANN inference for computation of momentum fluxes

* remove trailing space

* Initial commit

* address Robert Hallberg code review

* Restore deafult value of ZB_SCALING coefficient

---------

Co-authored-by: Alistair Adcroft <Alistair.Adcroft@noaa.gov>
Co-authored-by: Alistair Adcroft <adcroft@users.noreply.github.com>
